README.md
# knife-bastion

[![Code Climate](https://codeclimate.com/github/eligible/knife-bastion/badges/gpa.svg)](https://codeclimate.com/github/eligible/knife-bastion)
[![Gem Version](https://badge.fury.io/rb/knife-bastion.svg)](https://badge.fury.io/rb/knife-bastion)

This plugin allows Knife to access Chef server over a secure SSH connection,
without exposing Chef server port to your VPN network.

## Installation

If you're using ChefDK, install `knife-bastion` into the Chef omnibus:

```bash
chef gem install knife-bastion
```

If you're using rubygems to manage your Chef repository dependencies, add this line to your Chef repository's Gemfile:

```ruby
gem 'knife-bastion'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install knife-bastion

## Usage

Configure your bastion server in `.chef/knife.rb` (at the bottom):

```ruby
# ...
# your knife configurations goes here
# ...

# Bastion host SSH settings
knife[:bastion_host] = "bastion.mycorp.net"
knife[:bastion_user] = ENV["MYCORP_USER"] || ENV["CHEF_USER"] || ENV["USER"]
knife[:bastion_port] = "22"

# If you have multiple networks, that require different MFA tokens, specify
# each network name here. (This configuration is referenced to clarify the
# token a user should employ.)
# knife[:bastion_network] = "mynet"

# By default, the proxy server is created on port 4443. You may configure the
# local bastion port here:
# knife[:bastion_local_port] = 4443

require "knife-bastion/activate"
```

Now, your workflow will look like this:

1. Run `knife bastion start` - this command will establish SSH connection to
   bastion box for 10 minutes, and create a SOCKS proxy on port `4443`, that
   will forward all Chef requests to Chef server via bastion box.
2. Use Chef to do your work.
3. At any time you can use `knife bastion status` - which will verify the proxy
   and make sure everything works as expected.
4. After you finished, run `knife bastion stop` to shutdown the connection
   and turn off the proxy. If you forget to do this, it will die automatically
   after 10 minutes.

Sometimes when you work on a big change, default timeout of 10 minutes is too short.
You can increase timeout with `--timeout` flag:

```
knife bastion start --timeout 1800
```

Maximum timeout is 3600 (1 hour) for security reasons. You can re-establish bastion
connection by executing `knife bastion start` (if the connection is currently active,
it will be forcibly closed.)

### Bastion troubleshooting

If something is not right, you need to ensure you have access to bastion box.
Try connecting to `bastion.mycorp.net` via SSH:

```bash
ssh ${MYCORP_USER-$USER}@bastion.mycorp.net
```

Check current bastion connection status (it will tell you if there is anything
wrong with your box):

```
knife bastion status
```

### Using knife-bastion Outside of Chef

This gem can be used to connect to any server utilizing generic
proxy client. The bastion should be started the same way as it is
described above.

Example:

```ruby
require 'knife-bastion/client_proxy'
require 'rest_client'

# initialize RestClient::Resource object and wrap it into a bastion proxy
api = KnifeBastion::ClientProxy.new(
  RestClient::Resource.new('https://domain.com/api/v1')
)

# use api the same way as it would be used as a RestClient::Resource
api['posts/1/comments'].post 'Good article.', content_type: 'text/plain'
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

Bug reports and pull requests are welcome on GitHub at https://github.com/eligible/knife-bastion.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
## 1.2.2 (Development)

Features:
  - Make bastion port configurable 
  - Supports Ruby 3.1

## 1.2.1 (Development)

Features:
  - Updates for chef client compatability

## 1.2.0 (Development)

Features:
  - Berkshelf now supports bastion (`berks upload`)

## 1.1.1 (September 27, 2016)

Bugfixes:
  - Fixed the issue with `knife bastion status` plugin, when it sometimes failed to detect bastion host IP address

## 1.1.0 (August 30, 2016)

Changes:
  - Proxy code has been refactored to make it more generic, so it can be used to proxy any requests through bastion connection

Bugfixes:

## 1.0.0 (August 22, 2016)

Features:

  - Connect to bastion server via SSH and proxy all Chef requests through this connection
  - Knife plugins to monitor status, start and stop connections


lib/chef/knife/bastion_base.rb
require 'chef/knife'

class Chef
  class Knife
    class BastionBase < Knife
      include Chef::Mixin::ShellOut

      def initialize_params
        @bastion_user    = Chef::Config[:knife][:bastion_user] || ENV['CHEF_USER'] || ENV['USER']
        @bastion_host    = Chef::Config[:knife][:bastion_host]
        @bastion_network = Chef::Config[:knife][:bastion_network]
        @bastion_port    = Chef::Config[:knife][:bastion_port] || 22
        @chef_host       = URI.parse(Chef::Config[:chef_server_url]).host
        @local_port      = Chef::Config[:knife][:bastion_local_port] || 4443
      end

      def tunnel_pid(local_port, raise_on_closed_port = true)
        # Check if local port is open, get proxy process PID
        pid_result = shell_out("lsof -nPt -i4TCP:#{local_port} -sTCP:LISTEN")
        unless pid_result.status.success?
          if raise_on_closed_port
            ui.fatal "Tunnel is not open on port #{local_port}"
            abort
          end
          return nil
        end
        proxy_pid = pid_result.stdout.chomp

        # Verify tunnel destination
        bastion_ip_addr = Socket.getaddrinfo(@bastion_host, nil, :INET, :STREAM, Socket::IPPROTO_TCP).first[3]
        dest_result = shell_out("lsof -an -p #{proxy_pid} -i4@#{bastion_ip_addr}:#{@bastion_port}")
        unless dest_result.status.success?
          ui.fatal "There is a process with PID #{proxy_pid} listening on port #{local_port}, but it does not look like a tunnel"
          abort
        end

        proxy_pid
      end

      def print_tunnel_info(header, timeout: nil, pid: nil)
        ui.info <<-INFO
#{header}
  * Bastion host: #{ui.color "#{@bastion_user}@#{@bastion_host}", :bold, :white}
  *    Chef host: #{ui.color @chef_host, :bold, :white}
  *   Local port: #{ui.color @local_port.to_s, :bold, :white}
        INFO
        if timeout
          ui.info <<-INFO
  *      Timeout: #{ui.color timeout.to_s, :bold, :white} seconds
          INFO
        end
        if pid
          ui.info <<-INFO
  *    Proxy PID: #{ui.color pid.to_s, :bold, :white}
          INFO
        end
      end

      def run
        initialize_params

        # Retrieve proxy process PID. Raises an error if something is wrong
        proxy_pid = tunnel_pid(@local_port)
        print_tunnel_info("Found an esablished tunnel:", pid: proxy_pid)

        # This line will raise an exception if tunnel is broken
        response = check_tunnel_status('/policies')            
        if response.is_a?(Net::HTTPSuccess)
          ui.info ui.color("OK:  ", :green) + "The tunnel is up and running"
        else
          raise "Error: Tunnel is broken"
        end
      end

      def check_tunnel_status(path)
        require 'socksify/http'

        uri = URI.parse("https://#{@chef_host}#{path}")
        resp = nil
      
        Net::HTTP.socks_proxy('127.0.0.1', @local_port).start(uri.host, uri.port, use_ssl: (uri.scheme == 'https')) do |http|
          req = Net::HTTP::Get.new(uri.request_uri)
          resp = http.request(req)
        end
        resp
      end  
    end
  end
end
